src/server/middleware/api/github.js

```diff
@@ -5,61 +5,113 @@ const { isProduction } = require('../../utils')
+                  }
+                }
+              }
+            }
+          }
+        }
+      `,
+      {
+        headers: {
+          authorization: `token ${process.env.GITHUB_TOKEN}`
+        }
+      }
+    )
+
+    const issues = repository.issues.edges.map(({ node }) => ({
+      title: node.title,
+      url: node.url,
+      comments: node.comments.totalCount,
+      date: node.createdAt
+    }))
+
+    const stars = repository.stargazers.totalCount
+
+    cache.set('issues', issues)
+    cache.set('stars', stars)
+
+    return { issues, stars }
+  } catch (e) {
+    return { error: e.message || e }
+  }
+}
+
+async function issues(_, res) {
   if (!process.env.GITHUB_TOKEN) {
-    res.status(200).json({ issues: [] })
+    res
+      .status(200)
+      .json({ issues: [], error: 'No GITHUB_TOKEN specified on the server!' })
   } else {
     if (cache.get('issues')) {
       if (!isProduction) console.log('Using cache for "issues"')
 
       res.status(200).json({ issues: cache.get('issues') })
     } else {
       console.log('Not using cache for "issues"')
+      const { issues } = await getFreshGithubData()
+      if (!issues) {
+        res.status(404)
+      } else {
+        res.status(200).json({ issues })
+      }
     }
+  }
+}
 
-    try {
-      const { repository } = await graphql(
-        `
-          {
-            repository(owner: "iterative", name: "dvc") {
-              issues(
-                first: 3
-                states: OPEN
-                orderBy: { field: CREATED_AT, direction: DESC }
-              ) {
-                edges {
-                  node {
-                    title
-                    createdAt
-                    url
-                    comments {
-                      totalCount
-                    }
-                  }
-                }
-              }
-            }
-          }
-        `,
-        {
-          headers: {
-            authorization: `token ${process.env.GITHUB_TOKEN}`
-          }
-        }
-      )
-
-      const issues = repository.issues.edges.map(({ node }) => ({
-        title: node.title,
-        url: node.url,
-        comments: node.comments.totalCount,
-        date: node.createdAt
-      }))
-
-      cache.set('issues', issues)
+async function getStars() {
+  const cachedValue = cache.get('stars')
+  if (cachedValue) {
+    if (!isProduction) console.log('Using cache for "stars"')
+    return cachedValue
+  } else {
+    console.log('Not using cache for "stars"')
+    const { stars } = await getFreshGithubData()
+    return stars
+  }
+}
 
-      res.status(200).json({ issues })
-    } catch (e) {
+async function stars(req, res) {
+  if (!process.env.GITHUB_TOKEN) {
+    res.status(403).json({ error: 'No GITHUB_TOKEN specified on the server!' })
+  } else {
+    const stars = await getStars()
+    if (stars) {
+      // If the data in unchanged, return a minimal response indicating so.
+      if (Number(req.query.current) === stars) {
```

Comment:
feels like a premature optimization ... we save like a few bytes I suppose?

Reply:
I suppose you're right for stars. Something like this would serve better in an API passes around more data.